go/host/enclave/service.go

```diff
@@ -200,27 +175,108 @@ func (e *Service) Unsubscribe(id rpc.ID) error {
 	return e.GetEnclaveClient().Unsubscribe(id)
 }
 
-// promoteNewActiveSequencer is a background goroutine that promotes a new active sequencer at startup or when the current one fails.
-// It will never give up, it just cycles through current enclaves until one can be successfully promoted.
-func (e *Service) promoteNewActiveSequencer() {
-	for e.activeSequencerID.Load() == _noActiveSequencer && e.running.Load() {
-		for _, guardian := range e.enclaveGuardians {
-			enclID := guardian.GetEnclaveID()
-			e.logger.Info("Attempting to promote new sequencer.", log.EnclaveIDKey, enclID)
-			err := guardian.PromoteToActiveSequencer()
+// managePeriodicBatches is a background goroutine that triggers periodic batch production and ensures there is always
+// exactly one active sequencer enclave for that.
+// It will promote a new active sequencer if the current one fails or is unavailable.
+func (e *Service) managePeriodicBatches() {
+	interval := e.batchInterval
+	if interval == 0 {
+		interval = 1 * time.Second
+	}
+	// we use ticks rathen than sleeps to maintain batch production cadence
+	batchProductionTicker := time.NewTicker(interval)
+
+	for e.running.Load() {
+		// wait for the next tick
+		<-batchProductionTicker.C
+
+		activeSeq, err := e.getActiveSequencerGuardian()
+		if err != nil {
+			e.logger.Info("No active sequencer found, trying to promote a new one", log.ErrKey, err)
+			e.tryPromoteNewSequencer()
+			continue
+		}
+
+		if activeSeq.InSyncWithL1() {
+			err = activeSeq.ProduceBatch()
 			if err != nil {
-				e.logger.Info("Unable to promote new sequencer.", log.EnclaveIDKey, enclID, log.ErrKey, err)
+				// todo: do we want to have a few failed attempts before looking to promote a new one?
+				e.logger.Error("Sequencer failed to produce batch", log.ErrKey, err)
+				// if the active sequencer fails, we will try to promote a new one
+				e.tryPromoteNewSequencer()
 				continue
 			}
-			e.activeSequencerID.Store(enclID)
-			e.logger.Warn("Successfully promoted new sequencer.", log.EnclaveIDKey, enclID)
-			return
 		}
-		// wait for retry interval before trying again, enclaves may not be ready yet or may be awaiting permissioning
-		time.Sleep(_promoteSeqRetryInterval)
+
+		// sanity check: make sure there are no enclaves, which think they are active sequencers
+		for _, guardian := range e.enclaveGuardians {
```

Comment:
shouldn't this sanity check be the first thing ?

Reply:
In the case where just the host has restarted, when it comes back then one of the enclaves will still think it is active. If we did this check first then we would shut it down, but this way we will continue with it as the active sequencer. Maybe I should make that clearer in a comment.

This check is more of a clean-up and it's not about the risk of the bug we had (now that batch production is just centralised in one place in this service).